keps/sig-storage/3476-volume-group-snapshot/README.md

```diff
@@ -0,0 +1,1653 @@
+        VolumeGroupSnapshotName *string
+	......
+}
+
+type VolumeSnapshotContentStatus struct{
+	......
+        // +optional
+        VolumeGroupSnapshotContentName *string
+	......
+}
+```
+
+### Example Yaml Files
+
+#### Create VolumeGroupSnapshot
+
+Create a VolumeGroupSnapshotClass:
+```
+apiVersion: volumegroup.storage.k8s.io/v1alpha1
+kind: VolumeGroupSnapshotClass
+metadata:
+  name: volumeGroupSnapshotClass1
+spec:
+  parameters:
+     …...
+```
+
+A VolumeGroupSnapshot taken from multiple volumes dynamically:
+```
+apiVersion: volumegroup.storage.k8s.io/v1alpha1
+kind: VolumeGroupSnapshot
+metadata:
+  name: my-group-snapshot
+spec:
+  selector:
+    myapp: postgresql
+  volumeGroupSnapshotClassName: volumeGroupSnapshotClass1
+```
+
+The new VolumeGroupSnapshot logic will be added to the Snapshot Controller and the csi-snapshotter sidecar to handle VolumeGroupSnapshotClass, VolumeGroupSnapshot, and VolumeGroupSnapshotContent resources accordingly.
+
+Snapshot controller will also be modified so that it will not delete an indiviual VolumeSnapshot that is part of a VolumeGroupSnapshot. External snapshotter sidecar will be modified so that it will not delete an individual VolumeSnapshotContent that is part of a VolumeGroupSnapshotContent.
+
+### CSI Changes
+
+#### CSI Capabilities
+
+A new group controller service will be added with a new controller capability CREATE_DELETE_GET_VOLUME_GROUP_SNAPSHOT.
+
+* CREATE_DELETE_GET_VOLUME_GROUP_SNAPSHOT:
+  Indicates that the controller plugin supports creating, deleting, and getting details of a snapshot of
+  multiple volumes.
+
+#### CSI Group Controller RPC
+
+```
+service Controller {
+  …
+  rpc CreateVolumeGroupSnapshot(CreateVolumeGroupSnapshotRequest)
+    returns (CreateVolumeGroupSnapshotResponse) {
+        option (alpha_method) = true;
+    }
+
+  rpc DeleteVolumeGroupSnapshot(DeleteVolumeGroupSnapshotRequest)
+    returns (DeleteVolumeGroupSnapshotResponse) {
+        option (alpha_method) = true;
+    }
+
+  rpc GetVolumeGroupSnapshot(GetVolumeGroupSnapshotRequest)
+    returns (GetVolumeGroupSnapshotResponse) {
+        option (alpha_method) = true;
+    }
+  …
+}
+```
+
+#### CreateVolumeGroupSnapshot
+
+The purpose of this call is to request the creation of a multi-volume snapshot. Group snapshots can be created from multiple volumes. Note that calls to this function must be idempotent - the function may be called multiple times for the same name - the group snapshot must only be created once.
+
+```
+message CreateVolumeGroupSnapshotRequest {
+  option (alpha_message) = true;
+
+  // The suggested name for the group snapshot. This field is REQUIRED
+  // for idempotency.
+  // Any Unicode string that conforms to the length limit is allowed
+  // except those containing the following banned characters:
+  // U+0000-U+0008, U+000B, U+000C, U+000E-U+001F, U+007F-U+009F.
+  // (These are control characters other than commonly used whitespace.)
+  string name = 1;
```

Comment:
What is the name Kubernetes uses when it calls this?

Reply:
We'll follow what we did for the name in CreateSnapshot.  So it will be:
<Prefix>-<VolumeGroupSnapshot UUID>
Prefix is "snap" by default and it is configurable.
There is also a UUID length that is configurable.